### PR TITLE
refactor error handling so not all exceptions are caught

### PR DESCRIPTION
**Pull request by rbren**
I've been seeing reports where a code-level error gets caught by the `try/except`, and gets passed back to the agent as an `AgentErrorObservation`, even though there's nothing it can do to recover

This PR tries to be more selective about which types of errors we handle and pass back to the agent.

### CODE REVIEW

File: opendevin/controller/agent_controller.py [hunk 1]
```diff
@@ -170,21 +167,10 @@ async def step(self, i: int):
         try:
             action = self.agent.step(self.state)
             if action is None:
-                raise AgentNoActionError()
-            logger.info(action, extra={'msg_type': 'ACTION'})
-        except Exception as e:
+                raise AgentNoActionError('No action was returned')
+        except (AgentMalformedActionError, AgentNoActionError) as e:
             observation = AgentErrorObservation(str(e))
```
Comment (enyst) [inline opendevin/controller/agent_controller.py:172]: If I understand correctly, it's this kind of code, initializing the observation with str(e)? An alternative could be `AgentErrorObservation('Malformed response')`

File: opendevin/controller/agent_controller.py [hunk 2]
```diff
@@ -170,21 +167,10 @@ async def step(self, i: int):
...
-            logger.debug(traceback.format_exc())
-
-            # raise specific exceptions that need to be handled outside
-            # note: we are using classes from openai rather than litellm because:
-            # 1) litellm.exceptions.AuthenticationError is a subclass of openai.AuthenticationError
-            # 2) embeddings call, initiated by llama-index, has no wrapper for errors.
-            #    This means we have to catch individual authentication errors
-            #    from different providers, and OpenAI is one of these.
-            if isinstance(e, (AuthenticationError, ContextWindowExceededError, APIConnectionError)):
-                raise
```
Comment (enyst) [inline opendevin/controller/agent_controller.py:187]: I didn't test yet, afaik though, the exception `raised` here is the one caught in the caller to stop the loop, when there's no good api key or other unrecoverable errors. Is the intention to no longer stop the loop for unrecoverable errors?

File: opendevin/controller/agent_controller.py (hunk 1 above)
Comment (rbren) [reply opendevin/controller/agent_controller.py:172]: We create `AgentMalformedActionErrors` with a few different messages (e.g. "unsupported action type `foo`, action must be one of `read`, `write`, ..."). So `str(e)` should have more information

File: opendevin/controller/agent_controller.py (hunk 2 above)
Comment (rbren) [reply opendevin/controller/agent_controller.py:187]: The big change is that now errors get `raised` _by default_! We only catch a specific set of errors (MalformedAction, NoAction), and the rest are treated as unrecoverable and stop the loop.

**Comment by enyst** [comment]
I think I love these changes. 😅 

I also think we now need to send specifically all JSON errors to the LLM.

I went through everything I could find that is affected, and I think it should work well, but I'm less sure about the other agents than monologue. I suppose there is an inherent risk in that we really didn't have to catch errors we knew were handled downstream, as it is the case with JSON parsing. But it looks like the rest will never happen(tm) 😆

For the loops PR, I think it has to be even more simplified. 'Any AgentErrorObservation' now affects too much, etc.

**Review by enyst — APPROVED** [review]

**Pull request merged**